internal/resource/config/datastore/custom_data_store.go

```diff
@@ -0,0 +1,222 @@
+			Default:     objectdefault.StaticValue(types.ObjectNull(resourcelink.AttrType())),
+			Attributes:  resourcelink.ToSchema(),
+		},
+		"configuration": pluginconfiguration.ToSchema(),
+	}
+	customDataStoreSchema.Validators = []validator.Object{
+		objectvalidator.ExactlyOneOf(
+			path.MatchRelative().AtParent().AtName("jdbc_data_store"),
+			path.MatchRelative().AtParent().AtName("ldap_data_store"),
+			path.MatchRelative().AtParent().AtName("ping_one_ldap_gateway_data_store"),
+		),
+	}
+
+	return customDataStoreSchema
+}
+
+func toStateCustomDataStore(con context.Context, clientValue *client.DataStoreAggregation, plan basetypes.ObjectValue) (types.Object, diag.Diagnostics) {
+	var diags, allDiags diag.Diagnostics
+	customDataStore := *clientValue.CustomDataStore
+	pluginDescriptorRef, diags := resourcelink.ToState(con, &customDataStore.PluginDescriptorRef)
+	allDiags = append(allDiags, diags...)
+	parentRef, diags := resourcelink.ToState(con, customDataStore.ParentRef)
+	allDiags = append(allDiags, diags...)
+	configurationObject := func() (types.Object, diag.Diagnostics) {
+		planConfiguration, ok := plan.Attributes()["configuration"]
+		if ok {
+			return pluginconfiguration.ToState(planConfiguration.(types.Object), &customDataStore.Configuration)
+		} else {
+			return pluginconfiguration.ToState(types.ObjectNull(pluginconfiguration.AttrType()), &customDataStore.Configuration)
+		}
+	}
+	configurationToState, diags := configurationObject()
+	allDiags = append(allDiags, diags...)
+	customDataStoreVal := map[string]attr.Value{
+		"type":                  types.StringValue(customDataStore.Type),
+		"name":                  types.StringValue(customDataStore.Name),
+		"plugin_descriptor_ref": pluginDescriptorRef,
+		"parent_ref":            parentRef,
+		"configuration":         configurationToState,
+	}
+	customDataStoreObj, diags := types.ObjectValue(customDataStoreAttrType, customDataStoreVal)
+	allDiags = append(allDiags, diags...)
+	return customDataStoreObj, allDiags
+}
+
+func readCustomDataStoreResponse(ctx context.Context, r *client.DataStoreAggregation, state *dataStoreResourceModel, plan *types.Object) diag.Diagnostics {
+	var diags diag.Diagnostics
+	state.Id = types.StringPointerValue(r.CustomDataStore.Id)
+	state.CustomId = types.StringPointerValue(r.CustomDataStore.Id)
+	state.MaskAttributeValues = types.BoolPointerValue(r.CustomDataStore.MaskAttributeValues)
+	state.JdbcDataStore = jdbcDataStoreEmptyStateObj
+	state.CustomDataStore, diags = toStateCustomDataStore(ctx, r, *plan)
+	state.LdapDataStore = ldapDataStoreEmptyStateObj
+	state.PingOneLdapGatewayDataStore = pingOneLdapGatewayDataStoreEmptyStateObj
+	return diags
+}
+
+func addOptionalCustomDataStoreFields(addRequest client.DataStoreAggregation, con context.Context, createJdbcDataStore client.CustomDataStore, plan dataStoreResourceModel) error {
```

Comment:
nit: probably should be "createCustomDataStore" rather than "createJdbcDataStore" for the variable name